### PR TITLE
Adds Caching for Windows Feature Powershell provider

### PR DESCRIPTION
**Pull request by jakauppila**
### Description

Adds Caching for Windows Feature Powershell provider. Thanks @tas50 for doing all the heavy lifting

- I'm sure there's more efficient ways to doing a few things since we're dealing with json. It was more a straight port of what was done on the dism side
- This cut our subsequent runs from ~41 seconds to ~1.7 seconds for the ~30 features we install

Checks source definition key in registry when determining feature availability

### Check List

- [ ] All tests pass. See <https://github.com/chef-cookbooks/community_cookbook_documentation/blob/master/TESTING.MD>
- [ ] New functionality includes testing.
- [ ] New functionality has been documented in the README if applicable
- [x] All commits have been signed for the Developer Certificate of Origin. See <https://github.com/chef-cookbooks/community_cookbook_documentation/blob/master/CONTRIBUTING.MD>

### CODE REVIEW

**Comment by jakauppila** [comment]
So that failure confuses me. Is it not coercing the value correctly into an array or does it just write it out oddly?
```
windows_feature_powershell("Web-Asp-Net45, Web-Net-Ext45") do
  action [:install]
  default_guard_interpreter :default
  declared_type :windows_feature_powershell
  cookbook_name "test"
  recipe_name "feature"
  feature_name ["Web-Asp-Net45, Web-Net-Ext45"]
end
```

It would appear that it's not properly getting the difference between the two arrays at

https://github.com/chef-cookbooks/windows/pull/538/commits/21f32831b1244face392b426ceab837e922f54f5#diff-d10018fdcbb8eea540d70ca7f60e2031R132

File: resources/feature_dism.rb [hunk 1]
```diff
@@ -159,6 +159,9 @@ def add_to_feature_mash(feature_type, feature_string)
   # @return [void]
   def fail_if_removed
     return if new_resource.source # if someone provides a source then all is well
+    if node['os_version'].to_f > 6.2
+      return if registry_key_exists?('HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Policies\Servicing') && registry_value_exists?('HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Policies\Servicing', name: 'LocalSourcePath') # if source is defined in the registry, still fine
```
Comment (tas50) [inline resources/feature_dism.rb]: We should probably add a log message here in debug level

**Comment by tas50** [comment]
I'm getting this error when I try to run the feature test suite locally:

```
         * windows_feature[Remove FTP for Appveyor] action remove
           * windows_feature_powershell[Remove FTP for Appveyor] action remove

             ================================================================================
             Error executing action `remove` on resource 'windows_feature_powershell[Remove FTP for Appveyor]'
             ================================================================================

             JSON::ParserError
             -----------------
             765: unexpected token at ''
```

**Comment by jakauppila** [comment]
Anything particular about your environment in which it's failing? That's working fine for me locally and in Appveyor.

**Comment by jakauppila** [comment]
Been trying to get the array input to work as well, not getting very far. Looks like it affects dism too
`windows_feature_dism ['Web-Asp-Net45', 'Web-Net-Ext45']`

**Comment by jakauppila** [comment]
@tas50 Any thoughts on the failure to install via an array?

**Comment by tas50** [comment]
Let me pull it local and see what I can figure out

File: resources/feature_powershell.rb [hunk 2]
```diff
@@ -28,36 +28,60 @@
...
 
 action :install do
-  Chef::Log.warn("Requested feature #{new_resource.feature_name.join(',')} is not available on this system.") unless available?
-  unless !available? || installed?
-    converge_by("install Windows feature#{'s' if new_resource.feature_name.count > 1} #{new_resource.feature_name.join(',')}") do
+  reload_cached_powershell_data unless node['powershell_features_cache']
+  fail_if_unavailable # fail if the features don't exist
+  fail_if_removed # fail if the features are in removed state
+
+  reload_cached_powershell_data unless node['powershell_features_cache']
```
Comment (tas50) [inline resources/feature_powershell.rb]: The reload is in here twice

Comment (jakauppila) [reply resources/feature_powershell.rb]: Removed that duplicate

**Comment by jakauppila** [comment]
Added a test to install an array of features with dism.

From what I can tell, it's failing on this line when an array is supplied 

https://github.com/chef-cookbooks/windows/blob/master/resources/feature_dism.rb#L124

**Comment by tas50** [comment]
I have this all checked out locally and I'm working on a few things. I'll have something for you tomorrow

**Comment by tas50** [comment]
Here's the branch I'm working on https://github.com/chef-cookbooks/windows/commits/jakauppila-powershell_caching

**Comment by tas50** [comment]
Thanks for noticing the array functionality was broken and for putting this PR together. I'll get a PR open to add it to chef-client so we can ship this in Chef 14. It should seriously speed up windows runs.

**Comment by jakauppila** [comment]
Thanks for digging into it!

**Comment by tas50** [comment]
I put the tests into master as well. Thanks for those

**Pull request closed without merging**